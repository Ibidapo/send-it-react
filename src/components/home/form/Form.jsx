import React, { Component } from 'react';
import { isEmail } from 'validator';
import PropTypes from 'prop-types';
import { toast } from 'react-toastify';

import toastOptions from '../../../utils/toastOptions';
import './styles.scss';

class Form extends Component {
  state = {
    email: "",
    password: "",
    disableButton: false,
    displaySpinner: false,
  }

  onEmailChange = ({ target: { value } }) => this.setState(() => ({ email: value }));
  onPasswordChange = ({ target: { value } }) => this.setState(() => ({ password: value }));

  onSubmit = (e) => {
    e.preventDefault();
    const { props: { onSubmit }, state: { email, password } } = this;

    if (!isEmail(email)) return toast.error('Email address is not valid', toastOptions);
    if (password.length < 4) 
      return toast.error('Password must be greater than 4 characters', toastOptions);

    onSubmit({ email, password});
  }

  render() {
    const { 
      state: { email, password }, props: { button_label },
      onEmailChange, onPasswordChange, onSubmit
    } = this;

    return (
      <form id="form" className="mx-auto" onSubmit={onSubmit}>
        <div className='pb-1'>
          <label htmlFor="form-email">Email</label>
          <input
            type="text"
            value={email}
            onChange={onEmailChange}
          />
        </div>
        <div className="py-1">
          <label htmlFor="form-pass">Password</label>
            <input
              type="password"
              value={password}
              onChange={onPasswordChange}
            />
        </div>
        <div className="text-center pt-1">
          <button id="form-btn">{button_label}</button>
        </div>  
      </form>
    );
  }
}

Form.propTypes = {
  onSubmit: PropTypes.func,
  button_label: PropTypes.string
}

export default Form;