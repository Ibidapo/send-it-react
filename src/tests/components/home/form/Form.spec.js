import React from 'react';
import { shallow } from 'enzyme';

import Form from '../../../../components/home/form/Form';

let wrapper;

describe('Test for Form Component', () => {
  it('Should render correctly', () => {
    wrapper = shallow(<Form />);
    expect(wrapper).toMatchSnapshot();
  })

  it('Should set email state on input change', () => {
    const value = 'foo@bar.com';

    wrapper.find('input').at(0).simulate('change', { target: { value } });
    expect(wrapper.state('email')).toBe(value);
  });

  it('Should set password state on input change', () => {
    const value = 'foobar';

    wrapper.find('input').at(1).simulate('change', { target: { value } });
    expect(wrapper.state('password')).toBe(value);
  });

  it('Should not submit form with invalid email', () => {
    const props = { onSubmit: jest.fn() }

    wrapper = shallow(<Form {...props} />)
    wrapper.find('input').at(0).simulate('change', { target: { value: 'foocom' } });
    wrapper.find('input').at(1).simulate('change', { target: { value: 'foobar' } });

    wrapper.find('form').simulate('submit', { preventDefault: () => { } });
  });

  it('Should not submit form with invalid password', () => {
    const props = { onSubmit: jest.fn() }

    wrapper = shallow(<Form {...props} />)
    wrapper.find('input').at(0).simulate('change', { target: { value: 'foo@bar.com' } });
    wrapper.find('input').at(1).simulate('change', { target: { value: '' } });
    wrapper.find('form').simulate('submit', { preventDefault: () => { } });
  });

  it('Should submit form', () => {
    const props = { onSubmit: jest.fn() }

    wrapper = shallow(<Form {...props} />)
    wrapper.find('input').at(0).simulate('change', { target: { value: 'foo@bar.com' } });
    wrapper.find('input').at(1).simulate('change', { target: { value: 'foobar' } });
    wrapper.find('form').simulate('submit', { preventDefault: () => { } });
  });
});