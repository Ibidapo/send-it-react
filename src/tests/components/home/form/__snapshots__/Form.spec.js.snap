// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for Form Component Should render correctly 1`] = `
<form
  className="mx-auto"
  id="form"
  onSubmit={[Function]}
>
  <div
    className="pb-1"
  >
    <label
      htmlFor="form-email"
    >
      Email
    </label>
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="py-1"
  >
    <label
      htmlFor="form-pass"
    >
      Password
    </label>
    <input
      onChange={[Function]}
      type="password"
      value=""
    />
  </div>
  <div
    className="text-center pt-1"
  >
    <button
      id="form-btn"
    />
  </div>
</form>
`;
